File: src/flourish/analysis/talents/keeper/sm_cooldown_reduction.py
from flourish.analysis.attributor import TalentAttributor
from flourish.models.events import CastEvent, HealEvent
from flourish.tracking.hot_tracker import HotTracker
from flourish.tracking.buff_tracker import BuffTracker

SWIFTMEND = 18562
WILD_GROWTH = 48438
EARLY_SPRING_NODE_ID = 94591
EARLY_SPRING_TALENT_ID = 117895  # WCL entryId (choice node vs Bounteous Bloom)
DRYADS_DANCE_NODE_ID = 109713
RENEWING_SURGE_NODE_ID = 82060

PROSPERITY_NODE_ID = 82079
PROSPERITY_TALENT_ID = 103136  # WCL entryId (choice node vs Verdant Infusion)

DRYAD_HEAL_SPELLS = {1264659, 1264664}
SPIRIT_THICKET_SPELL = 1264905

BASE_SM_CD_MS = 15000
BASE_WG_CD_MS = 10000
WG_4PC_REDUCTION_MS = 2000
RENEWING_SURGE_REDUCTION_AVG = 0.195
EARLY_SPRING_REDUCTION_MS = 1000
DRYADS_DANCE_SPEED_FACTOR = 1.25
ON_COOLDOWN_TOLERANCE_MS = 1500
DRYAD_GAP_THRESHOLD_MS = 2000


def compute_effective_cd(
    has_renewing_surge: bool,
    has_early_spring: bool,
    dryad_overlap_ms: float,
) -> float:
    """Compute effective SM cooldown in ms given active talents and Dryad overlap."""
    cd = float(BASE_SM_CD_MS)
    if has_renewing_surge:
        cd *= (1 - RENEWING_SURGE_REDUCTION_AVG)
    if has_early_spring:
        cd -= EARLY_SPRING_REDUCTION_MS
    if dryad_overlap_ms > 0:
        overlap = min(dryad_overlap_ms, cd)
        remaining = cd - overlap
        cd = remaining + overlap / DRYADS_DANCE_SPEED_FACTOR
    return cd


def compute_effective_wg_cd(
    has_early_spring: bool,
    has_4pc: bool,
) -> float:
    """Compute effective WG cooldown in ms."""
    cd = float(BASE_WG_CD_MS)
    if has_4pc:
        cd -= WG_4PC_REDUCTION_MS
    if has_early_spring:
        cd -= EARLY_SPRING_REDUCTION_MS
    return cd


class WgCooldownReductionAttributor(TalentAttributor):
    """Attributes healing value of Early Spring WG CD reduction.

    Tracks WG cast gaps, checks on-cooldown usage, and attributes
    a fraction of downstream GG healing."""

    name = "WG Cooldown Reduction"

    def __init__(self, downstream_attributors=None, has_4pc=True):
        super().__init__()
        self._wg_cast_timestamps: list[int] = []
        self._downstream = downstream_attributors or []
        self._has_4pc = has_4pc

    def is_selected(self) -> bool:
        if self.combatant_info is None:
            return True
        return (
            EARLY_SPRING_NODE_ID in self.combatant_info.talent_nodes
            and EARLY_SPRING_TALENT_ID in self.combatant_info.talent_ids
        )

    def process_event(self, event, hot_tracker, buff_tracker):
        # Track begincast (not cast) for WG since it has a cast time.
        # The CD starts when you press the button, not when the cast finishes.
        if isinstance(event, CastEvent) and event.type == "begincast" and event.ability_id == WILD_GROWTH:
            self._wg_cast_timestamps.append(event.timestamp)

    def process_heal(self, event, hot_tracker, buff_tracker) -> float:
        return 0.0

    def finalize(self) -> float:
        if len(self._wg_cast_timestamps) < 2:
            return 0.0

        has_early_spring = (
            self.has_talent(EARLY_SPRING_NODE_ID)
            and self.combatant_info is not None
            and EARLY_SPRING_TALENT_ID in self.combatant_info.talent_ids
        )

        unreduced_cd = compute_effective_wg_cd(
            has_early_spring=False,
            has_4pc=self._has_4pc,
        )

        reduced_cd = compute_effective_wg_cd(
            has_early_spring=has_early_spring,
            has_4pc=self._has_4pc,
        )

        total_ratio = 0.0
        total_casts = len(self._wg_cast_timestamps)

        for i in range(1, total_casts):
            gap = self._wg_cast_timestamps[i] - self._wg_cast_timestamps[i - 1]
            if gap <= reduced_cd + ON_COOLDOWN_TOLERANCE_MS:
                ratio = 1 - (reduced_cd / unreduced_cd)
                total_ratio += max(ratio, 0.0)

        if total_ratio == 0.0:
            return 0.0

        extra_cast_fraction = total_ratio / total_casts
        downstream_total = sum(attr.total_attributed for attr in self._downstream)

        return extra_cast_fraction * downstream_total


class SmCooldownReductionAttributor(TalentAttributor):
    """Attributes healing value of Early Spring + Dryad's Dance SM CD reduction.

    Computes effective CD per SM cast, checks if pressed on cooldown,
    and attributes a fraction of downstream healing (SotF, PotA, GG)."""

    name = "SM Cooldown Reduction"

    def __init__(self, downstream_attributors=None):
        super().__init__()
        self._sm_cast_timestamps: list[int] = []
        self._dryad_windows: list[tuple[int, int]] = []
        self._dryad_start: int | None = None
        self._dryad_last_heal: int | None = None
        self._downstream = downstream_attributors or []

    def is_selected(self) -> bool:
        if self.combatant_info is None:
            return True
        has_early_spring = (
            EARLY_SPRING_NODE_ID in self.combatant_info.talent_nodes
            and EARLY_SPRING_TALENT_ID in self.combatant_info.talent_ids
        )
        has_dryads_dance = DRYADS_DANCE_NODE_ID in self.combatant_info.talent_nodes
        return has_early_spring or has_dryads_dance

    def _close_dryad_window(self):
        if self._dryad_start is not None and self._dryad_last_heal is not None:
            self._dryad_windows.append((self._dryad_start, self._dryad_last_heal))
            self._dryad_start = None
            self._dryad_last_heal = None

    def process_event(self, event, hot_tracker: HotTracker, buff_tracker: BuffTracker):
        all_dryad_spells = DRYAD_HEAL_SPELLS | {SPIRIT_THICKET_SPELL}

        # Track Dryad windows from pet heal events
        if (
            isinstance(event, HealEvent)
            and event.ability_id in all_dryad_spells
            and self.combatant_info is not None
            and event.source_id != self.combatant_info.source_id
        ):
            # Close previous window if gap exceeded
            if (
                self._dryad_last_heal is not None
                and event.timestamp - self._dryad_last_heal > DRYAD_GAP_THRESHOLD_MS
            ):
                self._close_dryad_window()

            if self._dryad_start is None:
                self._dryad_start = event.timestamp
            self._dryad_last_heal = event.timestamp
        else:
            # Non-dryad event: check if open window should close due to gap
            if (
                self._dryad_last_heal is not None
                and event.timestamp - self._dryad_last_heal > DRYAD_GAP_THRESHOLD_MS
            ):
                self._close_dryad_window()

        if isinstance(event, CastEvent) and event.ability_id == SWIFTMEND:
            self._sm_cast_timestamps.append(event.timestamp)

    def process_heal(self, event, hot_tracker, buff_tracker) -> float:
        return 0.0

    def _dryad_overlap_in_window(self, window_start: float, window_duration: float) -> float:
        """Compute total Dryad overlap during a recharge window."""
        cd_end = window_start + window_duration
        overlap = 0.0
        for d_start, d_end in self._dryad_windows:
            o_start = max(window_start, d_start)
            o_end = min(cd_end, d_end)
            if o_end > o_start:
                overlap += o_end - o_start
        return overlap

    def finalize(self) -> float:
        self._close_dryad_window()

        if len(self._sm_cast_timestamps) < 2:
            return 0.0

        has_renewing_surge = self.has_talent(RENEWING_SURGE_NODE_ID)
        has_early_spring = (
            self.has_talent(EARLY_SPRING_NODE_ID)
            and self.combatant_info is not None
            and EARLY_SPRING_TALENT_ID in self.combatant_info.talent_ids
        )
        has_dryads_dance = self.has_talent(DRYADS_DANCE_NODE_ID)
        has_prosperity = (
            self.has_talent(PROSPERITY_NODE_ID)
            and self.combatant_info is not None
            and PROSPERITY_TALENT_ID in self.combatant_info.talent_ids
        )

        max_charges = 2 if has_prosperity else 1

        # Unreduced CD = baseline with Renewing Surge only (no Early Spring, no Dryad)
        unreduced_cd = compute_effective_cd(
            has_renewing_surge=has_renewing_surge,
            has_early_spring=False,
            dryad_overlap_ms=0,
        )

        # Charge simulation: track pending recharges as (completion_time, reduced_cd)
        charges = max_charges
        pending: list[tuple[float, float]] = []

        total_ratio = 0.0
        total_casts = len(self._sm_cast_timestamps)

        for cast_ts in self._sm_cast_timestamps:
            # Restore charges that have finished recharging
            was_depleted = charges == 0
            last_restore: tuple[float, float] | None = None

            while pending and pending[0][0] <= cast_ts:
                entry = pending.pop(0)
                charges = min(charges + 1, max_charges)
                if was_depleted:
                    last_restore = entry
                    was_depleted = charges == 0

            on_cooldown = False
            reduced_cd_used = 0.0

            if charges == 0 and pending:
                # Still depleted — check if next charge arrives within tolerance
                completion, rcd = pending[0]
                if completion <= cast_ts + ON_COOLDOWN_TOLERANCE_MS:
                    on_cooldown = True
                    reduced_cd_used = rcd
                    pending.pop(0)
                    charges += 1
                else:
                    # Model imprecision: cast happened but no charge predicted.
                    # Force-restore without counting as on-cooldown.
                    pending.pop(0)
                    charges += 1
            elif last_restore is not None:
                # Was depleted, charge recently restored — check tolerance
                completion, rcd = last_restore
                if cast_ts - completion <= ON_COOLDOWN_TOLERANCE_MS:
                    on_cooldown = True
                    reduced_cd_used = rcd

            charges -= 1

            # Schedule recharge for consumed charge
            recharge_start = pending[-1][0] if pending else cast_ts
            dryad_overlap = (
                self._dryad_overlap_in_window(recharge_start, unreduced_cd)
                if has_dryads_dance
                else 0.0
            )
            reduced_cd = compute_effective_cd(
                has_renewing_surge=has_renewing_surge,
                has_early_spring=has_early_spring,
                dryad_overlap_ms=dryad_overlap,
            )
            pending.append((recharge_start + reduced_cd, reduced_cd))

            if on_cooldown:
                ratio = 1 - (reduced_cd_used / unreduced_cd)
                total_ratio += max(ratio, 0.0)

        if total_ratio == 0.0:
            return 0.0

        extra_cast_fraction = total_ratio / total_casts

        # Sum downstream attributor totals
        downstream_total = sum(attr.total_attributed for attr in self._downstream)

        return extra_cast_fraction * downstream_total

File: go-backend/internal/talents/sm_cooldown_reduction.go
package talents

import (
	"math"

	"github.com/rdruid-talent-analyzer/go-backend/internal/models"
	"github.com/rdruid-talent-analyzer/go-backend/internal/tracking"
)

const (
	EarlySpringNodeID   = 94591
	EarlySpringTalentID = 117895
	DryadsDanceNodeID   = 109713
	RenewingSurgeNodeID = 82060
	ProsperityNodeID    = 82079
	ProsperityTalentID  = 103136

	baseSMCDMS               = 15000
	baseWGCDMS               = 10000
	wg4pcReductionMS         = 2000
	renewingSurgeReductionAvg = 0.195
	earlySpringReductionMS   = 1000
	dryadsDanceSpeedFactor   = 1.25
	onCooldownToleranceMS    = 1500
	dryadGapThresholdMS      = 2000
)

func ComputeEffectiveCd(hasRenewingSurge, hasEarlySpring bool, dryadOverlapMS float64) float64 {
	cd := float64(baseSMCDMS)
	if hasRenewingSurge {
		cd *= (1 - renewingSurgeReductionAvg)
	}
	if hasEarlySpring {
		cd -= earlySpringReductionMS
	}
	if dryadOverlapMS > 0 {
		overlap := math.Min(dryadOverlapMS, cd)
		remaining := cd - overlap
		cd = remaining + overlap/dryadsDanceSpeedFactor
	}
	return cd
}

func ComputeEffectiveWgCd(hasEarlySpring, has4pc bool) float64 {
	cd := float64(baseWGCDMS)
	if has4pc {
		cd -= wg4pcReductionMS
	}
	if hasEarlySpring {
		cd -= earlySpringReductionMS
	}
	return cd
}

// SmCooldownReductionAttributor tracks SM cast timestamps and Dryad windows.
type SmCooldownReductionAttributor struct {
	BaseAttributor
	smCasts        []int
	dryadWindowsVal [][2]int // {start, end}
	dryadStart     *int
	dryadLastHeal  *int
	downstream     []TalentAttributor
}

func NewSmCooldownReductionAttributor(downstream []TalentAttributor) *SmCooldownReductionAttributor {
	return &SmCooldownReductionAttributor{
		BaseAttributor: NewBaseAttributor("Early Spring + Dryad's Dance", nil, nil),
		downstream:     downstream,
	}
}

func (a *SmCooldownReductionAttributor) IsSelected() bool {
	if a.CombatantInfo == nil {
		return true
	}
	hasES := a.CombatantInfo.TalentNodes[EarlySpringNodeID] && a.CombatantInfo.TalentIDs[EarlySpringTalentID]
	hasDD := a.CombatantInfo.TalentNodes[DryadsDanceNodeID]
	return hasES || hasDD
}

func (a *SmCooldownReductionAttributor) SmCastTimestamps() []int { return a.smCasts }
func (a *SmCooldownReductionAttributor) DryadWindows() [][2]int  { return a.dryadWindowsVal }

func (a *SmCooldownReductionAttributor) closeDryadWindow() {
	if a.dryadStart != nil && a.dryadLastHeal != nil {
		a.dryadWindowsVal = append(a.dryadWindowsVal, [2]int{*a.dryadStart, *a.dryadLastHeal})
		a.dryadStart = nil
		a.dryadLastHeal = nil
	}
}

func isDryadHealSpell(id int) bool {
	return id == DryadTranquility || id == DryadRegrowthSpell || id == SpiritOfTheThicket
}

func (a *SmCooldownReductionAttributor) ProcessEvent(event models.Event, hot *tracking.HotTracker, buff *tracking.BuffTracker) {
	// Track Dryad windows from pet heal events
	if he, ok := event.(*models.HealEvent); ok && isDryadHealSpell(he.AbilityID) && a.IsPlayerPet(he.SourceID) {
		if a.dryadLastHeal != nil && he.Timestamp-*a.dryadLastHeal > dryadGapThresholdMS {
			a.closeDryadWindow()
		}
		if a.dryadStart == nil {
			ts := he.Timestamp
			a.dryadStart = &ts
		}
		ts := he.Timestamp
		a.dryadLastHeal = &ts
	} else {
		// Non-dryad event: close window if gap exceeded
		if a.dryadLastHeal != nil && event.GetBase().Timestamp-*a.dryadLastHeal > dryadGapThresholdMS {
			a.closeDryadWindow()
		}
	}

	if ce, ok := event.(*models.CastEvent); ok && ce.AbilityID == Swiftmend {
		a.smCasts = append(a.smCasts, ce.Timestamp)
	}
}

func (a *SmCooldownReductionAttributor) dryadOverlapInWindow(windowStart, windowDuration float64) float64 {
	cdEnd := windowStart + windowDuration
	overlap := 0.0
	for _, w := range a.dryadWindowsVal {
		oStart := math.Max(windowStart, float64(w[0]))
		oEnd := math.Min(cdEnd, float64(w[1]))
		if oEnd > oStart {
			overlap += oEnd - oStart
		}
	}
	return overlap
}

type pendingCharge struct {
	completion float64
	reducedCD  float64
}

func (a *SmCooldownReductionAttributor) Finalize() float64 {
	a.closeDryadWindow()

	if len(a.smCasts) < 2 {
		return 0.0
	}

	hasRS := a.HasTalent(RenewingSurgeNodeID)
	hasES := a.HasTalent(EarlySpringNodeID) && a.CombatantInfo != nil && a.CombatantInfo.TalentIDs[EarlySpringTalentID]
	hasDD := a.HasTalent(DryadsDanceNodeID)
	hasProsperity := a.HasTalent(ProsperityNodeID) && a.CombatantInfo != nil && a.CombatantInfo.TalentIDs[ProsperityTalentID]

	maxCharges := 1
	if hasProsperity {
		maxCharges = 2
	}

	unreducedCD := ComputeEffectiveCd(hasRS, false, 0)

	charges := maxCharges
	var pending []pendingCharge
	totalRatio := 0.0
	totalCasts := len(a.smCasts)

	for _, castTS := range a.smCasts {
		wasDepleted := charges == 0
		var lastRestore *pendingCharge

		for len(pending) > 0 && pending[0].completion <= float64(castTS) {
			entry := pending[0]
			pending = pending[1:]
			if charges < maxCharges {
				charges++
			}
			if wasDepleted {
				e := entry
				lastRestore = &e
				wasDepleted = charges == 0
			}
		}

		onCooldown := false
		reducedCDUsed := 0.0

		if charges == 0 && len(pending) > 0 {
			entry := pending[0]
			if entry.completion <= float64(castTS)+onCooldownToleranceMS {
				onCooldown = true
				reducedCDUsed = entry.reducedCD
				pending = pending[1:]
				charges++
			} else {
				pending = pending[1:]
				charges++
			}
		} else if lastRestore != nil {
			if float64(castTS)-lastRestore.completion <= onCooldownToleranceMS {
				onCooldown = true
				reducedCDUsed = lastRestore.reducedCD
			}
		}

		charges--

		rechargeStart := float64(castTS)
		if len(pending) > 0 {
			rechargeStart = pending[len(pending)-1].completion
		}
		dryadOverlap := 0.0
		if hasDD {
			dryadOverlap = a.dryadOverlapInWindow(rechargeStart, unreducedCD)
		}
		reducedCD := ComputeEffectiveCd(hasRS, hasES, dryadOverlap)
		pending = append(pending, pendingCharge{rechargeStart + reducedCD, reducedCD})

		if onCooldown {
			ratio := 1 - (reducedCDUsed / unreducedCD)
			totalRatio += math.Max(ratio, 0.0)
		}
	}

	if totalRatio == 0.0 {
		return 0.0
	}

	extraCastFraction := totalRatio / float64(totalCasts)
	downstreamTotal := 0.0
	for _, attr := range a.downstream {
		downstreamTotal += attr.GetTotalAttributed()
	}

	return extraCastFraction * downstreamTotal
}

// WgCooldownReductionAttributor tracks WG cast timestamps.
type WgCooldownReductionAttributor struct {
	BaseAttributor
	wgCasts    []int
	downstream []TalentAttributor
	has4pc     bool
}

func NewWgCooldownReductionAttributor(downstream []TalentAttributor, has4pc bool) *WgCooldownReductionAttributor {
	return &WgCooldownReductionAttributor{
		BaseAttributor: NewBaseAttributor("Early Spring (WG)", nil, nil),
		downstream:     downstream,
		has4pc:         has4pc,
	}
}

func (a *WgCooldownReductionAttributor) IsSelected() bool {
	if a.CombatantInfo == nil {
		return true
	}
	return a.CombatantInfo.TalentNodes[EarlySpringNodeID] && a.CombatantInfo.TalentIDs[EarlySpringTalentID]
}

func (a *WgCooldownReductionAttributor) WgCastTimestamps() []int { return a.wgCasts }

func (a *WgCooldownReductionAttributor) ProcessEvent(event models.Event, hot *tracking.HotTracker, buff *tracking.BuffTracker) {
	if ce, ok := event.(*models.CastEvent); ok && ce.Type == models.EventBeginCast && ce.AbilityID == WildGrowth {
		a.wgCasts = append(a.wgCasts, ce.Timestamp)
	}
}

func (a *WgCooldownReductionAttributor) Finalize() float64 {
	if len(a.wgCasts) < 2 {
		return 0.0
	}

	hasES := a.HasTalent(EarlySpringNodeID) && a.CombatantInfo != nil && a.CombatantInfo.TalentIDs[EarlySpringTalentID]

	unreducedCD := ComputeEffectiveWgCd(false, a.has4pc)
	reducedCD := ComputeEffectiveWgCd(hasES, a.has4pc)

	totalRatio := 0.0
	totalCasts := len(a.wgCasts)

	for i := 1; i < totalCasts; i++ {
		gap := a.wgCasts[i] - a.wgCasts[i-1]
		if float64(gap) <= reducedCD+onCooldownToleranceMS {
			ratio := 1 - (reducedCD / unreducedCD)
			totalRatio += math.Max(ratio, 0.0)
		}
	}

	if totalRatio == 0.0 {
		return 0.0
	}

	extraCastFraction := totalRatio / float64(totalCasts)
	downstreamTotal := 0.0
	for _, attr := range a.downstream {
		downstreamTotal += attr.GetTotalAttributed()
	}

	return extraCastFraction * downstreamTotal
}
